win_xml module for manipulating XML files on Windows

richardcs opened the pull request:
##### SUMMARY
New module to handle XML files on windows

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
win_xml

##### ANSIBLE VERSION

```
ansible --version
ansible 2.4.0 (devel 0a6dcf4169) last updated 2017/07/03 20:14:36 (GMT -700)
  config file = None
  configured module search path = [u'/home/richardl/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/richardl/ansible/lib/ansible
  executable location = /home/richardl/ansible/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```


##### ADDITIONAL INFORMATION
win_lineinfile is adequate for properties types of files but XML is harder to deal with. Since Powershell has native XML handling, this module just uses that for manipulating XML.

dagwieers commented:
@richardcs I noticed you implemented the `backup:` parameter, which is mostly missing from all the other Windows modules. We definitely need to port this to the other modules (and maybe move this functionality to the powershell.ps1 library).

Are you willing to look at this and propose an implementation for the other modules (win_copy, win_template, win_lineinfile, ...) ?

If so, I would like to add it to the action-plan and agenda for next meeting.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+    attribute:
+        description:
+        - "The attribute name IFF type is 'attribute'"
+        required: no
```
If a parameter is not required, you don't have to add this. This is implied, `required:` is only needed for parameters that are required.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+        - text
+    attribute:
+        description:
+        - "The attribute name IFF type is 'attribute'"
```
You don't have to quote that string. The general rule is to only quote in YAML where it is needed.
(So people grasp how YAML works from our documents, over time...)

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+    backup:
+        description:
+        - Whether to backup the remote server's XML before applying the change (yes/no)
+        required: false
```
Not required ? Don't say so. This is implied.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+    root:
+        description:
+        - The root of the remote server XML where the fragment will go
+        required: false
```
Not required ? Don't say so. This is implied.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+version_added: "2.4"
+short_description: Add XML fragment to an XML parent
+description:
+    - Adds XML fragments formatted as strings to existing XML on remote servers
```
The description needs to have a trailing dot. (The short_description does not)

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+options:
+    path:
+        description:
+        - The path of remote servers XML
```
Same here, trailing do required.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+        aliases: [ xpath ]
+    backup:
+        description:
+        - Whether to backup the remote server's XML before applying the change (yes/no)
```
Don't add (yes/no), that's generated by the documentation automatically.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+        description:
+        - Whether to backup the remote server's XML before applying the change (yes/no)
+        required: false
+        default: no
```
For booleans, you also need to add:
```yaml
    type: bool
```

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+        - "The attribute name IFF type is 'attribute'"
+        required: no
+
+author: "Richard Levenberg (richard.levenberg@cosocloud.com)"
```
Do this as a list, unquoted:
```yaml
author:
- Richard Levenberg (@richardcs)
```

Your email can go at the top of the file for the copyright.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+        default: no
+    type:
+        description:
+        - The type of XML you are working with
```
Trailing dot.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+EXAMPLES = r'''
+# Apply our filter to Tomcat web.xml
+- win_xml:
+   path: C:\\apache-tomcat\webapps\myapp\WEB-INF\web.xml
```
The double backslashes are not needed here. They are actually unwanted.

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,95 @@
+
+# Apply sslEnabledProtocols to Tomcat's server.xml
+- win_xml:
+   path: C:\\Tomcat\conf\server.xml
```
The double backslashes are not needed here. They are actually unwanted.

dagwieers commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,191 @@
+$params = Parse-Args $args -supports_check_mode $true
+$dest = Get-AnsibleParam $params "path" -type "path" -FailIfEmpty $true -aliases "dest", "file"
+$fragment = Get-AnsibleParam $params "fragment" -type "str" -FailIfEmpty $true -aliases "xmlstring"
+$root = Get-AnsibleParam $params "root" -type "str" -FailIfEmpty $false -Default "DocumentElement" -aliases "xpath"
```
Don't add `-failifempty $false`, this is implied. This option is only needed if it is `$true`.

dagwieers commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,191 @@
+$type = Get-AnsibleParam $params "type" -type "str" -FailIfEmpty $false -Default "element" -ValidateSet "element", "attribute", "text"
+$attribute = Get-AnsibleParam $params "attribute" -type "str" -FailIfEmpty ($type -eq "attribute")
+
+$result = New-Object PSObject @{
```
Please make this a normal hashtable, not a PSObject.

(And I wonder where you copied that from, because we are trying to kill all these wrong examples everywhere ;-))

dagwieers commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,191 @@
+$attribute = Get-AnsibleParam $params "attribute" -type "str" -FailIfEmpty ($type -eq "attribute")
+
+$result = New-Object PSObject @{
+    win_xml = New-Object PSObject
```
Please don't add the module-name as part of the result. This makes everything harder for users to use.
We put all the result data directly as part of the result hash table.

dagwieers commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,191 @@
+
+    if ($add) {
+        $result.changed = $TRUE
+        if (-Not $params._ansible_check_mode) {
```
Please add check_mode the same way we do for other modules, using Get-AnsibleParam.
Also, where possible, make use of `-WhatIf` support, and add `-WhatIf` support to the backup function, so we can use that like `-WhatIf:$check_mode` as we tend to do for most cmdlets.

dagwieers commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,191 @@
+	        $node.SetAttributeNode($attribute, $xmlnew.$root.NamespaceURI)
+            }
+            $node.SetAttribute($attribute, $fragment)
+            $xmlnew.Save("$dest")
```
Only the `$xmlnew.Save($dest)` needs to be skipped in check-mode. We really want to do as much as possible in check-mode except the change itself. So that if any condition, or problem is found in check-mode as well (usually before performing it in production). So the more it can tests, the better ! (That's also why we prefer `-WhatIf` support in cmdlets and functions, so we leave it up to the cmdlet how far it can take this)

dagwieers reviewed (changes requested):
Looking good overall. Some cosmetic changes, and some requests to put it more into line with other modules.

dagwieers commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,191 @@
+$params = Parse-Args $args -supports_check_mode $true
+$dest = Get-AnsibleParam $params "path" -type "path" -FailIfEmpty $true -aliases "dest", "file"
+$fragment = Get-AnsibleParam $params "fragment" -type "str" -FailIfEmpty $true -aliases "xmlstring"
+$root = Get-AnsibleParam $params "root" -type "str" -FailIfEmpty $false -Default "DocumentElement" -aliases "xpath"
```
I am surprised you did not call this `xpath`, root is not the correct name for this.

richardcs replied:
In the 'element' case, you are required to pass in a root element under which the element fragment will be placed. So its specifically not XPath. One option is to change the element handling to also use XPath instead of the powershell object. Both text and attribute are XPath, element is currently NOT XPath.

dagwieers replied:
Hmm, need to look how this is done with the xml module. But having it default to root will be very confusing for people that expected xpath (even when that is an alias).

Worst case, we need to add both parameters and add conditionals so people are using it correctly. Again, let's see first how this is one with the xml module...

Commit:
win_xml module for manipulating XML files on Windows

documentation fixups

handling backup in a more ansible canonical way

remove quotes from $dest

Handle elements with only text child nodes

dagwieers commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,94 @@
+    backup:
+        description:
+        - Whether to backup the remote server's XML before applying the change.
+        type: boolean
```
This should be `type: bool` IIRC.

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+$dest = Get-AnsibleParam $params "path" -type "path" -FailIfEmpty $true -aliases "dest", "file"
+$fragment = Get-AnsibleParam $params "fragment" -type "str" -FailIfEmpty $true -aliases "xmlstring"
+$xpath = Get-AnsibleParam $params "xpath" -type "str" -FailIfEmpty $true
+$backup = Get-AnsibleParam $params "backup" -type "bool" -Default "no" -ValidateSet "no", "yes"
```
This can be simplified to `$backup = Get-AnsibleParam $params -name "backup" -type "bool" -default $false`

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+                $add = $FALSE
+                break
+            } catch {
+                $_.InvocationInfo | Out-File c:\win_xml.log
```
If you wish to keep the errors for debugging purposes it is preferable to add it to the result object and not output to a file on the host. This could break if the user Ansible is running with doesn't have permissions in this directory.

jborean93 commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,94 @@
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'core'}
```
Can you change supported_by to community

jborean93 commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,94 @@
+        description:
+        - Whether to backup the remote server's XML before applying the change.
+        type: bool
+        default: no
```
this should be `default: 'no'` with the quotes

jborean93 commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,94 @@
+        - text
+    attribute:
+        description:
+        - The attribute name IFF type is 'attribute'.
```
You should also add another sentence saying this is required if `C(type=attribute)`

jborean93 commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,94 @@
+    - Richard Levenberg (@richardcs)
+'''
+
+RETURN = r'''
```
A few things on this section
* Are you able to move this after the examples like we have done with the other modules.
* You currently list only 1 return value when i see there are more being set in your code
* Don't worry about adding the `changed` return value as that is common with Ansible
* The documentation for `msg` below seems to be incorrect, are you able to update it to reflect what that variable represents

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+                throw "attribute value mismatch for actual=" + $actual.Name
+            }
+        }
+    }
```
You may also want to check the cases where actual has attributes while expected does not and vice versa

richardcs replied:
Lines 29-31 check the attribute counts so that covers the check as specified. I'm pretty sure the logic in lines 32-38 covers the checks to make sure that the child node is not there. The issue might be complex if a similar child is there, then what is the intent?

jborean93 replied:
You are only checking the count when both actual and expected have attributes. What if actual has attributes but expected does not, in this case there is a mismatch with the xml but it isn't picked up here.

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+            try {
+                Compare-XmlDocs $xmlchild.DocumentElement $element
+                $add = $FALSE
+                break
```
I might be missing something here, why are we breaking the loop after checking the first element, shouldn't `Compare-XmlDocs` be run on all the elements?

richardcs replied:
The logic here is that if any child of the original matches our fragment to add, we don't want to add the new fragment, and we don't care about other children whether they match or not.

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+    if ($add) {
+        $result.changed = $TRUE
+        if (-Not $check_mode) {
+            if ($backup -eq "yes") {
```
Now that `$backup` is a boolean, this should be `if ($backup -eq $true)` or just if `if ($backup)`

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+    if ($add) {
+        $result.changed = $TRUE
+        if (-Not $check_mode) {
+            if ($backup -eq "yes") {
```
Now that `$backup` is a boolean, this should be `if ($backup -eq $true)` or just if `if ($backup)`

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+                $result.backup = BackupFile($dest)
+            }
+            $node.InnerText = $fragment
+            $xmlnew.Save("$dest")
```
I don't think you need to quote `$dest`

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+    if ($add) {
+        $result.changed = $TRUE
+        if (-Not $check_mode) {
+            if ($backup -eq "yes") {
```
Now that `$backup` is a boolean, this should be `if ($backup -eq $true)` or just if `if ($backup)`

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,182 @@
+	        $node.SetAttributeNode($attribute, $xmlnew.DocumentElement.NamespaceURI)
+            }
+            $node.SetAttribute($attribute, $fragment)
+            $xmlnew.Save("$dest")
```
quotes around `$dest`

jborean93 reviewed (changes requested):
Thanks for the PR @richardcs I've added some comments to the PR. A few things that I wish to note overall
* Are you able to add some tests for this module, let us know if you need some direction on this
* It would be good to have some try/catch handlers on functions where you are saving the XML so we can set better error messages

dagwieers commented:
Ideally, **win_xml** and **xml** would be using the same test examples :-)

dagwieers commented:
+label new_module

trondhindenes commented:
Great stuff! I just have one comment (that's bitten me a few times): xml will from time to time care about the order of items - it would be great to somehow implement support for that in this module. Possibly an optional "after" param which would take an xpath for matching, and then insert the new item right after the matched one.

dagwieers commented:
@trondhindenes Could you check the **xml** module ? We would like to have the **win_xml** module to be  as close to the **xml** module as possible. https://github.com/ansible/ansible/pull/25323

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,194 @@
+    Fail-Json $result "Specified path $dest does not exist or is not a file."
+}
+
+[xml]$xmlorig = Get-Content -Path $dest
```
Would be good to put a try catch around this and return a more descriptive error like
`Fail-Json $result "failed to parse file at '$dest' as an XML document: $($_.Exception.Message)"`

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,194 @@
+$params = Parse-Args $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+$debug_level = Get-AnsibleParam -obj $params -name "_ansible_verbosity" -type "int"
```
Is this being used for anything?

richardcs replied:
This is used to determine the debug level See below.

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,194 @@
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
+$debug_level = Get-AnsibleParam -obj $params -name "_ansible_verbosity" -type "int"
+$debug = $debug_level -gt 2
```
Same as above, is this being used for anything. I'm also not sure what the Ansible practice is around returning extra values based on the verbosity level set.

richardcs replied:
The $debug variable is used to determine whether to return XML comparison errors. I can take all the functionality out or take the debug gate out. This was work in concert with removing the emission of the errors to a log on the remote server.

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,194 @@
+                break
+            } catch {
+                if ($debug) {
+                    result.err += $_.InvocationInfo
```
You would have to set `$result.err = @()` before doing this as it doesn't know err is an array.

richardcs replied:
I believe it does exactly this on lines 122-124

jborean93 commented on lib/ansible/modules/windows/win_xml.py:
```diff
@@ -0,0 +1,94 @@
+   type: attribute
+'''
+
+RETURN = r'''
```
Are you also able to add the `backup` and `err` return values and a description for those.

jborean93 commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,194 @@
+[xml]$xmlorig = Get-Content -Path $dest
+$xmlnew = $xmlorig.Clone()
+if ($type -eq "element") {
+    $xmlchild = [xml]$fragment
```
Another try/catch here to detect if it is a valid xml fragment and return a more descriptive message if not.

mattclay commented:
CI failure in integration test `windows: win_xml : check element add result that=[u'not element_add_result_second|changed']
`:

```
 {
    "assertion": "not element_add_result_second|changed", 
    "changed": false, 
    "evaluated_to": false, 
    "failed": true
}
```

richardcs commented:
@trondhindenes could you provide a use case for your example. If the XPATH is selecting a root element with a bunch of child nodes, are you suggesting another XPATH that selects one of the child nodes and inserts the fragment after that selected child node? For a first pass at a module that seems difficult from a conceptual standpoint (explaining to others that functionality) but if its a common use case out in the world I would happy to support that for the 'element' type.

richardcs commented:
@trondhindenes  Could you provide a use-case for your example mentioned above. Are you saying that the XPATH would select a root node with many child nodes, then the 'after' parameter would select a child node under that root, then the fragment would be added after that child? For a first pass at a module this seems complex from a conceptual point of view (explaining it to others). If its a common use case, I would be happy to implement it for the 'element' type.

jborean93 commented:
@richardcs I'm fairly happy with it so far and am looking to merge it in soon. The last thing that would be good to have is to use the same tests that https://github.com/ansible/ansible/pull/25323 has. It tests a few more use cases and will definitely help with regression testing in the future. You can see these tests at https://github.com/dagwieers/ansible/tree/a1ae785c5b9aec8f9a862f7a994fc7b736dae181/test/integration/targets/xml

dagwieers commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,226 @@
+
+$params = Parse-Args $args -supports_check_mode $true
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+
```
Having diff support would be nice.
I plan to add it once **xml** is merged upstream.

dagwieers commented:
So the module freeze deadline for v2.4 is getting closer (2017-08-29). Now is the time to finish up, get it reviewed and merged with no delay.

mattclay commented:
CI failure in integration test `test/integration/targets/win_xml/tasks/main.yml:43 / [windows-2012-R2_RTM] windows: win_xml : check element add result that=[u'not element_add_result_second|changed']`:

```
{
    "assertion": "not element_add_result_second|changed", 
    "changed": false, 
    "evaluated_to": false, 
    "failed": true
}
```

dagwieers commented:
@richardcs The xml module has been merged now. What is your plan to get this into v2.4 ?

mattclay commented:
CI failure in integration test `test/integration/targets/win_xml/tasks/main.yml:43 / [windows-2012-R2_RTM] windows: win_xml : check element add result that=[u'not element_add_result_second|changed']`:

```
{
    "assertion": "not element_add_result_second|changed", 
    "changed": false, 
    "evaluated_to": false, 
    "failed": true
}
```

trondhindenes commented:
@richardcs an example that's bitten me is the "Advanced Logging Module in IIS, the order of items here control the order in which items are written to the IIS log. So, changing the order of these fields will have an effect:
```xml
<logDefinition rollLogFileOnConfigChanges="true" writeLogDataToDisk="true" baseFileName="rikstv-advancedlogging" logRaiseTraceEvents="false" enabled="true" logRollOption="Schedule" maxDurationSeconds="86400" maxFileSizeKB="1024" schedule="Daily" publishLogEvent="false">
    <selectedFields>
        <logField id="Date-UTC" logHeaderName="date" required="false" defaultValue="" />
        <logField id="Time-UTC" logHeaderName="time" required="false" defaultValue="" />
        <logField id="Server Name" logHeaderName="s-computername" required="false" defaultValue="" />
        <logField id="Server-IP" logHeaderName="s-ip" required="false" defaultValue="" />
        <logField id="Method" logHeaderName="cs-method" required="false" defaultValue="" />
```

How to do this right is not something we have solved in-house yet, so it's just an ask more than a requirement. Don't let this stop you guys from merging.

dagwieers commented:
@richardcs I updated the **xml** integration tests to no longer rely on the diff command. So we can adapt the *test/integration/targets/xml* tests to work with **win_xml** as well.

dagwieers commented:
@richardcs Good news x2
- I have added diff support to the **xml** module, and I think you should do the same in **win_xml**, this greatly improves troubleshooting https://github.com/ansible/ansible/pull/28592
- We are going to add a new **diff** task directive, so we can mark all our tests to show diff output by default. This will help troubleshooting if integration tests fail https://github.com/ansible/ansible/pull/28581

JonPaz commented:
Hi there all,

Was just wondering what is happening wrt this module as it is something that I would find beneficial and could help to push things along a bit.

dagwieers commented:
Beware that the deadline for getting new features/modules accepted in [Ansible v2.5 is nearing](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_5.rst), it is set to either 2018-01-15 or 2018-01-31. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to [the Windows Working Group meeting agenda](https://github.com/ansible/community/issues/195) to get it resolved.

abhijeetka commented on lib/ansible/modules/windows/win_xml.ps1:
```diff
@@ -0,0 +1,241 @@
+            [void]$node.AppendChild($child)
+            $result.msg = "xml added"
+            $changed = $true
+        } elseif ($present -and ($state -eq "absent")) {
```
@richardcs I am trying to remove an element from xml file but its not working for me. I am trying following, Please correct me if i am doing wrong.

`- name: Running XML Transformations with state Absent
     win_xml:
       path:      "path\to\xml\file"
       xpath:     "/configuration/configSections/sectionGroup/section[@name='authentication']"
       state:     "absent"
       fragment: " "`

petemounce replied:
Should the xpath be `//configuration/configSections/sectionGroup/section[@name='authentication']` - double initial slash?

abhijeetka replied:
@petemounce it worked for other paramerters without adding double initial slash. its not working only if i am trying to delete an element.

richardcs replied:
I dont think you can have an empty fragment. The xpath should find the root of your XML where the child fragment is, then it will remove that child fragment.

So to guess maybe your xpath is "/configuration/configSections/sectionGroup/" and your fragment is "section[@name='authentication']"

richardcs replied:
@petemounce  "//" means find it at the root, which is probably artificially the case, but not explicitly necessary .

petemounce replied:
Ah, pardon.

abhijeetka replied:
@richardcs it didn't worked for me , if we give any value in fragment, it tries to add that element( even with the combination of state: "absent" ).

`-name: Running XML Transformations with state Absent
   win_xml: 
       path: "path\to\xml\file" 
       xpath: "/configuration/configSections/sectionGroup/" 
       state: "absent" 
       fragment: "section[@name='authentication']"
`
it gives following error : **Failed to parse fragment as XML: Cannot convert value \"section[@name='authorization']\" to type \"System.Xml.XmlDocument\". Error: \"The specified node cannot be inserted as the valid child of this node, because the specified node is the wrong type.\**

richardcs replied:
Could you provide the XML or a facsimile thereof? I I have enough info to reproduce and test what you are trying to do but need some XML to test with.

abhijeetka replied:
```
<?xml version="1.0" encoding="utf-8"?>
<!--
  For more information on how to configure your ASP.NET application, please visit
  http://go.microsoft.com/fwlink/?LinkId=169433
  -->
<configuration>
  <configSections>
      <sectionGroup name="applicationSettings" type="System.Configuration.ApplicationSettingsGroup, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" >
          <section name="GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings" type="System.Configuration.ClientSettingsSection, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" requirePermission="false" />
      </sectionGroup>
	 <sectionGroup name="GKE.ds.security" type="GKE.DS.Security.Configuration.SecuritySectionGroup, GKE.DS.Security">
      <section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />
      <section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />    
     </sectionGroup>
	<!--
  Adding a multiline comment
  second line comment
  third line
  -->
  </configSections>
  <system.web>
    <compilation debug="true" targetFramework="4.5" />
    <httpRuntime targetFramework="4.5" />
  </system.web>
  <appSettings>
  </appSettings>
<applicationSettings>
    <GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
      <setting name="Setting1" serializeAs="String">
        <value>Setting1Value</value>
      </setting>
    </GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
  </applicationSettings>
</configuration>
```

abhijeetka replied:
We need to remove these lines.
`<section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />
      <section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />    `

richardcs replied:
I apologize, I was confused. If you were adding an element you would specify the XPath and the entire element you want added for the fragment. I think for removing you would also have to specify the entire fragment you want removed. The module recursively checks for "equality" so you  have to have the entire fragment. Also fragments are not treated as XPath, sorry for that mistake. 

So your xpath would be "/configuration/configSections" (n.b. no ending slash) and you would need to use the module twice, once with fragment as `'<section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />'` and again with fragment as `'<section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />'`

abhijeetka replied:
This also didn't worked for me.

`-name: Running XML Transformations with state Absent win_xml: path: "path\to\xml\file" xpath: "/configuration/configSections" state: "absent" fragment: "<section name='authorization' type='GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security' />"`

this passes the task but nothing changes. I am completely stuck here, I need to find a way to delete the elements from the xml file.

Do you have any example for which the "absent" keyword works?

richardcs replied:
Ok, I have tested this and verified that this works. The way GitHub formats the code blocks makes it hard to read the XML you're working with. The XPath should be `'/configuration/configSections/sectionGroup[@name="GKE.ds.security"]'` and the full playbook used is as follows:

```
---
 - name: win_xml module test with absent
   hosts: []
   tasks:
     - name: run the win_xml module
       win_xml:
          path: C:\\Users\richardcs\Temp\test.xml
          xpath: '/configuration/configSections/sectionGroup[@name="GKE.ds.security"]'
          fragment: '<section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />'
          type: element
          state: absent
```

abhijeetka replied:
Okay I will give it a try.

abhijeetka replied:
@richardcs its worked ..!! thanks for helping me out. In Fragment , you have to give element exactly the same as in your xml file.

richardcs replied:
@abhijeetka yes, the logic there is if you were adding it, the default behavior, you would have to give the entire XML fragment you wanted to add, obviously. So to delete, you do the same. Glad it worked for you.

abhijeetka replied:
@richardcs one last thing that i want to know, Can we add the capability to delete even the attributes from the elements?
We are trying to remove only the attributes from the xml element.
For. Ex `<language default="eng" style="arial"></language>`
We want to delete the default attribute only.

richardcs replied:
I don't think the current module supports absent for attributes. If you give me the .xml you are working with I can probably add it.

abhijeetka replied:
@richardcs   Here is the xml file.
```
<?xml version="1.0" encoding="utf-8"?>
<!--
  For more information on how to configure your ASP.NET application, please visit
  http://go.microsoft.com/fwlink/?LinkId=169433
  -->
<configuration>
  <configSections>
      <sectionGroup name="applicationSettings" type="System.Configuration.ApplicationSettingsGroup, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" >
          <section name="GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings" type="System.Configuration.ClientSettingsSection, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089" requirePermission="false" />
      </sectionGroup>
	 <sectionGroup name="GKE.ds.security" type="GKE.DS.Security.Configuration.SecuritySectionGroup, GKE.DS.Security">
      <section name="authorization" type="GKE.DS.Security.Configuration.AuthorizationSection, GKE.DS.Security" />
      <section name="authentication" type="GKE.DS.Security.Configuration.AuthenticationSection, GKE.DS.Security" />    
     </sectionGroup>
	<!--
  Adding a multiline comment
  second line comment
  third line
  -->
  </configSections>
  <system.web>
    <compilation debug="true" targetFramework="4.5" />
    <httpRuntime targetFramework="4.5" />
  </system.web>
  <appSettings>
  </appSettings>
<applicationSettings>
    <GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
      <setting name="Setting1" serializeAs="String">
        <value>Setting1Value</value>
      </setting>
    </GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings>
  </applicationSettings>
</configuration>
```
we want to remove a attribute `serializeAs="String"` which is present in  `<setting name="Setting1" serializeAs="String">` element.

richardcs replied:
@abhijeetka I modified the win_xml module to support your use case. It was tested with this playbook. 
```
---
 - name: win_xml module test with absent attribute
   hosts: []
   tasks:
     - name: run the win_xml module
       win_xml:
          path: C:\\Users\richardcs\Temp\test.xml
          xpath: '/configuration/applicationSettings/GKE.DS.ALM.Patterns.DotNet.App.Web.Properties.Settings/setting'
          type: attribute
          fragment: String
          attribute: serializeAs
          state: absent
```

abhijeetka replied:
@richardcs  that works like a charm ...!! Thanks for adding this functionality.

dagwieers commented:
Beware that the deadline for getting new features/modules accepted in [Ansible v2.7 is nearing](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_7.rst#release-schedule), it is set to either 2018-08-23 or 2018-08-30. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to the [Windows Working Group meeting agenda](https://github.com/ansible/community/issues/294) to get it resolved.
